test_data/rules/prefer_final_parameters.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) 2021, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+// test w/ `dart test -N prefer_final_parameters`
+
+void badRequiredPositional(String label) { // LINT
+  print(label);
+}
+
+void goodRequiredPositional(final String label) { // OK
+  print(label);
+}
+
+void badOptionalPosition([String? label]) { // LINT
+  print(label);
+}
+
+void goodOptionalPosition([final String? label]) { // OK
+  print(label);
+}
+
+void badRequiredNamed({required String label}) { // LINT
+  print(label);
+}
+
+void goodRequiredNamed({required final String label}) { // OK
+  print(label);
+}
+
+void badOptionalNamed({String? label}) { // LINT
+  print(label);
+}
+
+void goodOptionalNamed({final String? label}) { // OK
+  print(label);
+}
+
+void badExpression(int value) => print(value); // LINT
+
+void goodExpression(final int value) => print(value); // OK
+
+bool _testingVariable;
+
+void set badSet(bool setting) => _testingVariable = setting; // LINT
+
+void set goodSet(final bool setting) => _testingVariable = setting; // OK
+
+var badClosure = (Object random) { // LINT
+  print(random);
+};
+
+var goodClosure = (final Object random) { // OK
```

Comment:
Sorry I didn't think of this on the first pass, but what about a case like
```
var uncertainClosure = (parameterWithNoType) => 0;
```
Such closures are common when passed as arguments to another function. Should those also be `final`? I believe that they will be linted, and I'm guessing that's the style you want to enforce. (A test isn't necessary, I mostly want to know that this is intentional.)

Reply:
Great question.  If the answer is no (and maybe either way) it might be good to update the docs to weigh in.

Reply:
You are correct in that they will currently be linted, but do you think that this will discourage people from using the lint as a whole?

I am worried that this structure so common, especially in functional like methods like the ones present in `Stream` and `Iterable` where you end up with multiple calls in a row that accept closures, usually not modifying anything anyway.

Perhaps in this case(where the parameter type can be assumed?) we shouldn't trigger the lint or even remove all the closure cases from this lint introduce a separate one specific to closures?

Let me know what you both think, thanks!

Reply:
I don't know for sure, but I do know that when you wrote this lint you didn't make any of the parameters final anywhere. Are you sure this is your preferred style? :-)

As for closures, would you have wanted to have written
```
parameters.parameters.forEach((final e) {
```

Reply:
Since we went with the `var` locals in linter, I tried my best to stick with that for parameters as well.

Truly, I would prefer to not write `final e` and instead have that be the default for Dart.

Thinking about it though, while I haven't really used `final` in my closures much, I think it makes sense to do so if I am making the rest of my parameters `final` as well. Especially since the `prefer_final_locals` will trigger in the closure too. The lint is for those who want that extra static guarantee, so I think it fits for those who will use it.